### ISSUE TITLE
Use generic AST provider to support non-Lisps

### ISSUE THREAD

**Issue by countvajhula** [countvajhula/symex.el#3]
At the moment symex uses paredit and lispy to derive the underlying tree representation to operate on. These packages are specific to Lisp syntax, whereas tree-structured data comes in many forms besides just symbolic expressions. Using an AST provider (such as tree-sitter: https://github.com/ubolonton/emacs-tree-sitter/) would extend the applicability of the symex language to non-Lisp languages like python, and also to data representations like XML.

Additional context: https://www.reddit.com/r/emacs/comments/ciswlw/symex_mode_editing_lisp_code_in_a_vimlike_way/evaq0hi/
#2

**Comment by tommy-mor** [comment]
Any update on this thought? Emacs-tree-sitter has come a very long way since you made this issue, and I[ was messing with this concept](https://github.com/tommy-mor/treenav) (built from the ground up with just emacs-tree-sitter) and it seems very achievable.

**Comment by countvajhula** [comment]
@tommy-mor I haven't had a chance to look at tree-sitter again, although I've definitely had it on the radar. That concept you've got going looks intriguing! From a quick try, I see this error: `Symbol's value as variable is void: tmy-cursor-mode`. I'll take a closer look when I get some time, but in the meantime, if you want to play with it some more, here are some thoughts regarding what it would look like to get symex to use tree-sitter or some other AST provider.

Symex essentially has 3 layers:
1. The modal interface
2. The symex DSL
3. The primitive operations

So for instance if we wanted to experiment with Evil instead of hydra, that would mean replacing (1) but largely leaving the other two intact. And if we want to add the AST, that would mean replacing (3) and leaving the other two intact. I'm not sure exactly how well-separated these layers are, but as I recall, at some point in the development I was intentional about keeping them separated so it's probable that most of the changes would just be replacing all of the implementations of the functions in [symex-primitives.el](https://github.com/countvajhula/symex.el/blob/master/symex-primitives.el) and also [symex-transformations.el](https://github.com/countvajhula/symex.el/blob/master/symex-transformations.el) with new ones that use tree-sitter. The former contains movement-related primitives, while the latter contains the transformations (like delete, emit/"barf", capture/"slurp", etc.).

Replacing the movement-related primitives is likely the low hanging fruit here, and your concept seems to target this. As far as replacing the transformations, we'd be talking about something like reimplementing paredit on top of tree-sitter, and I wonder if maybe that's best attempted as an independent project altogether, which is simply leveraged from symex.

**Comment by tommy-mor** [comment]
separate project definitely is good idea, if it can't fit nicely into symex. cool thing is though, if symex worked with tree sitter, it would also work with lisp because tree sitter could support lisp eventually (if it didn't already)

**Comment by countvajhula** [comment]
Yeah, symex as a language-agnostic structural editing DSL would be pretty compelling. Let's see!

**Comment by polaris64** [comment]
Now that there's increasing interest in integrating tree-sitter with Emacs (see [this Reddit post](https://www.reddit.com/r/emacs/comments/pxpq8d/rfc_emacs_treesitter_integration/) for example), has there been any more thought on this? I'm very excited by the possibilities!

**Comment by polaris64** [comment]
@countvajhula thank you for all of the information on the current progress.

I'm fairly new to tree sitter's API (and Symex too for that matter), so I've been using @ubolonton's [elisp-tree-sitter](https://github.com/emacs-tree-sitter/elisp-tree-sitter) to experiment. I've been creating simple tree traversal functions (just next/prev sibling and parent as a start) so see how that could be integrated as part of Symex's motions.

These are just personal experiments for now, but I'll be sure to let you know if anything more concrete comes of it.  I'll be sure to check out @tommy-mor's proof-of-concept too as I feel he's probably doing something very similar :)

**Comment by countvajhula** [comment]
Great, let me know how it goes! I'm happy to answer any questions.

**Comment by polaris64** [comment]
Hi @countvajhula,

I've spent today experimenting and I've created the following proof-of-concept: https://github.com/polaris64/symex-ts 

It only really supports navigation at the moment, but I've tested it in buffers containing Python and C source code and it seems to work quite nicely. As a next step I'd like to provide more functionality for editing (moving a block, capturing the next statement into an `if` body, etc.) but this is a necessary first step.

I'm using a hydra for the interface due to its simplicity. I think that if/when the concept has been proven then that's the point at which my project ends its life, and it could be incorporated into Symex.el if it's of any use to you. Therefore I think at this stage a hydra interface is sufficient.

If you are able to take a look then I'd be happy to hear your feedback!

**Comment by countvajhula** [comment]
Nice @polaris64! I'll take a look soon.

**Comment by tommy-mor** [comment]
I tried it, much better than my version :). \ One thing I am not sure about is in python (and probably other languages too) the grammar is deep enough where different levels are not necessary visually distinct.

For example entering function bodies with a single statement, you usually get one ``block`` object, the you have to call ``symex-ts-move-child`` one more time in order to get the underlying statement. (and the selection bounds stay the same)

In my [code](https://github.com/tommy-mor/treenav/blob/345c249e3ef2fdd74368426b6fb9295c60704fbd/treenav.el#L22), I designed the movements to keep running until the actual selection changed points, so that every movement is guaranteed to move the selection because pressing a movement key and staying in the same place is jarring.

Another option  would be to always skip past ``block`` groups in python, because afaik most of the cases I saw with this, block was the culprit. This would not be language agnostic though.

**Comment by polaris64** [comment]
Hi @tommy-mor,

I'm sure it's no better, just a different attempt at doing the same thing :) the more attempts the better, especially if it raises useful suggestions like yours has.

I had exactly the same thought as you regarding the `block` nodes. That's why I added highlighting to it, as I was getting confused when that happened!

This is actually my second attempt; my first attempt was too naive as it used the point as the only state. This seemed cleaner initially, however it caused issues with node structures like these. So my second version keeps track of the current node and that seems to work better.

For situations like those with `block`, your suggestion makes sense. I think that descending should bypass all nodes that contain no siblings at that level. As you say, it's a little jarring when you perform an action and seemingly nothing happens. Hopefully doing it this way will keep it language-agnostic, but we'll see.

Thanks for testing it! :)

**Comment by countvajhula** [comment]
Thank you both for your efforts. I tried it out and here are some thoughts:

1. This is great! Initially I was getting some weird error resembling `wrong type argument: user-ptrp nil`. It turned out that this was because, amazingly enough, tree-sitter doesn't support Elisp yet, or any lisp as far as I could tell.

So I wrote these wrapper functions for testing:

```
(defun symex--new-backward (&optional count)
  "Go backwards.

Use tree sitter if available, or fall back to lisp."
  (interactive)
  (if tree-sitter-mode
      (symex-ts-move-prev-sibling)
    (symex--backward count)))

(defun symex--new-forward (&optional count)
  "Go forward.

Use tree sitter if available, or fall back to lisp."
  (interactive)
  (if tree-sitter-mode
      (symex-ts-move-next-sibling)
    (symex--forward count)))

(defun symex--new-exit (&optional count)
  "Go down.

Use tree sitter if available, or fall back to lisp."
  (interactive)
  (if tree-sitter-mode
      (symex-ts-move-parent)
    (symex--exit count)))

(defun symex--new-enter (&optional count)
  "Go up.

Use tree sitter if available, or fall back to lisp."
  (interactive)
  (if tree-sitter-mode
      (symex-ts-move-child)
    (symex--enter count)))
```

... and modified [`symex--execute-tree-move`](https://github.com/countvajhula/symex.el/blob/25c05c1d8c37c05a9c6807bb410a8ea51049d28c/symex-evaluator.el#L38-L59) to use these wrapper implementations:
```
    (cond ((> move-x 0)
           (symex--new-forward move-x))
          ((< move-x 0)
           (symex--new-backward (abs move-x)))
          ((> move-y 0)
           (symex--new-enter move-y))
          ((< move-y 0)
           (symex--new-exit (abs move-y))))))
```
... and this allowed traversals to work! E.g. `f` and `b` traversed expressions fully in python. This brings us to the issue with blocks:

3. I agree that something needs to be done about blocks. For a function like this one:

```
    def my_func(arg):
        arg
```

... there seemed to be 3(!) nested levels on `arg` in the body. Is it because it's treating it as "function body" -> "block" -> "statement" or something like that?

> For situations like those with block, your suggestion makes sense. I think that descending should bypass all nodes that contain no siblings at that level.

this sounds like a good way to do it.

4. Unrelated to tree-sitter, but these overlays are great! We should reimplement symex's [existing "highlight" feature](https://github.com/countvajhula/symex.el/blob/25c05c1d8c37c05a9c6807bb410a8ea51049d28c/symex-misc.el#L554-L555) with them, since at the moment it uses regions which tends to interfere with other functionality in Symex.

On next steps, I think it'd be great if we could get this prototype to the point where all the motions (e.g. in python) behave analogously to the way they do in Lisp buffers. This would probably involve:
- addressing blocks in a sane way
- using the wrapper/fallback behavior for the elementary motions in `symex--execute-tree-move` (like the example above).
- verifying the behavior of each motion, i.e. the ones in the [Movement](https://countvajhula.com/2021/09/25/the-animated-guide-to-symex/#ib-toc-anchor-1) section of the Animated Guide. From initial testing, I noticed `leap-branch` didn't seem to work, but I didn't investigate further. If this is the only one that doesn't work, well, it's not such a big deal. I wouldn't want you to get too much in the weeds on this unless you wanted to :)
- verifying that the tree-sitter motions support the `count` argument. This would probably just involve adding a `dotimes` if tree-sitter doesn't already support counts.
- I noticed that some commands like `o` (`open-line-below`) don't work (in python) because they [use `forward-sexp` directly](https://github.com/countvajhula/symex.el/blob/25c05c1d8c37c05a9c6807bb410a8ea51049d28c/symex-transformations.el#L334-L339), to go to the end of the expression before adding a newline. We could probably define a new primitive interface like `symex-end-of-exp` to do the same thing here but with the wrapper behavior above, where it tries to do it using tree-sitter and falls back to the lisp way. We'd probably need to try out each feature individually to see if they work, and if not, modify some primitive function calls in their implementation to use a tree-sitter based equivalent (and fall back to lisp). We can always add these over time, too, so again, the prototype could focus on motions if that helps keep the scope contained.
- it looks like calling `symex-ts-set-current-node-from-point` causes tree-sitter to rebuild/update the AST, which is convenient since we'd probably just need to wrap the existing call to `symex-select-nearest` with this [on symex state entry](https://github.com/countvajhula/symex.el/blob/25c05c1d8c37c05a9c6807bb410a8ea51049d28c/symex.el#L117) to ensure any fresh changes are already taken into account while we're in symex state.
- symex-minor-mode currently maps `(` and `)` [to the paredit behavior](https://github.com/countvajhula/symex.el/blob/25c05c1d8c37c05a9c6807bb410a8ea51049d28c/symex.el#L56-L92). This would probably need to be revisited for non-Lisps, e.g. it shouldn't add a space. Not a top priority though.

Finally, I must mention again that you guys are gonna be mostly on your own until November since I'll be traveling on vacation for a week in October and then will be doing a talk at RacketCon and have to prepare for that 😅. Still, if you find yourself getting lost in the weeds on anything I'll do my best to make myself available for questions 👍

**Comment by countvajhula** [comment]
I went ahead and created [a new branch](https://github.com/countvajhula/symex.el/tree/symex-ts-integration) that uses symex-ts. This just fits the symex-ts module into Symex as a new primitives provider, used alongside the former primitives layer which is now treated as lisp-specific. It should make it easier to work on the prototype incrementally and see how it interacts with other symex features.

To use it, assuming you're using straight.el, you could put the relevant config in your init.d to point it at the `symex-ts-integration` branch on Github, or even download the repo locally. This is my symex/symex-ts config fwiw:
```
(use-package symex
  :after symex-ts
  :straight
  (symex
   :local-repo "~/.emacs.d/my-packages/symex"
   :type git)
  :custom
  (symex-modal-backend 'evil)
  (symex-quote-prefix-list (list "'" "`" "#'" "#`"))
  (symex-unquote-prefix-list (list "," ",@" "#,@"))
  :config
  (symex-initialize)
  (define-key
    symex-mode-map
    (kbd "C-w")
    (lambda ()
      (interactive)
      (execute-kbd-macro (kbd "(")))))

(use-package symex-ts
  :straight
  (symex-ts
   :local-repo "~/work/emacs/symex-ts"
   :type git))
```

**Comment by polaris64** [comment]
Hi @countvajhula,

Thanks for taking the time to look at it and to provide such thorough notes!

> It turned out that this was because, amazingly enough, tree-sitter doesn't support Elisp yet, or any lisp as far as I could tell.

Yes, that amused me when I noticed! I had a quick look and it looks like tree-sitter supports a generic "s-expression" grammar, so I'd imagine that'd work with Elisp. I'll give it a try.

> Unrelated to tree-sitter, but these overlays are great! We should reimplement symex's existing "highlight" feature with them, since at the moment it uses regions which tends to interfere with other functionality in Symex.

I'm glad that's useful, feel free to take any parts you feel are of benefit to Symex.

> Finally, I must mention again that you guys are gonna be mostly on your own until November since I'll be traveling on vacation for a week in October and then will be doing a talk at RacketCon and have to prepare for that

No problem, enjoy your vacation and good luck with the preparations!

Thanks too for adding a new branch for this development, I was going to ask if that would be possible. I also agree with how you've structured it; I imagined these new tree-sitter-based functions as being treated as a "provider" for the rest of the Symex code, so what you've done seems to mesh nicely.

I too am a little busy at present, but I'll be sure to update this thread whenever I can.

**Comment by countvajhula** [comment]
@polaris64 Do you think the movement-related code is ready to be merged into Symex? If so, do you want to create a PR for that? It could just include the work as it currently is (assuming it is in a functional state), with symex-ts.el as a new module, featuring just the motions initially. We can add additional functionality over time, including deletion and things like that. This way, your commits would be against the main repo and it could be more streamlined with e.g. forks and pull requests for each new set of features, etc.

WDYT?

**Comment by polaris64** [comment]
Hi @countvajhula,

I don't think we're too far off! I'm quite happy with how all of the simple movements are working, however I think there's more work to be done to handle the more complex movements.

For example, when I try to "leap" or "soar", I get the following error: `symex-ts--set-current-node: Lisp nesting exceeds ‘max-lisp-eval-depth’`. I haven't had chance to debug it yet, but it seems like an infinite loop to me.

Hopefully I'll be able to debug and fix those issues soon though, at which point I think it would be a good idea to merge it to Symex as you suggested. I'll keep you posted!

**Comment by countvajhula** [comment]
@polaris64 I took a quick look and there were a few things I noticed but I couldn't be sure if it was just mismatched configuration. So I'd say go ahead and create the PR against the `symex-ts-integration` branch, adding the new symex-ts.el module. I can then take a closer look and make more specific comments on the actual PR to keep things in context.

**Comment by SignSpice** [comment]
I am so glad to see this!  

I found this project recently -- might be helpful as a reference: https://github.com/ethan-leba/tree-edit

**Comment by countvajhula** [comment]
tree-edit certainly looks interesting, and there may be scope for integrating with it in some way down the line. For now, there is work underway at #33 thanks to the efforts of @polaris64 . I haven't found the time to continue the work yet as I wrap up some other things. But, I'm considering it a priority and hope to return to it soon!

**Comment by ethan-leba** [comment]
author of tree-edit here, if y'all are interested in using the tree-edit library i'd be happy to help or answer any questions! the library splits off the structural editing operations off from the evil package so i suspect it would slot in well with the code you currently have. that said, tree-edit is not very battle tested, but i think it's getting there.

**Comment by countvajhula** [comment]
Hi @ethan-leba ! Excellent, that is good to know.

To summarize the status of the work here: At the moment, due to the diligent efforts of @polaris64 , proof-of-concept tree-sitter integration is close to ready, and it follows Symex conventions as far as the user experience goes. Part of the work in #33 has been not just about tree-sitter but also making Symex tree-sitter ready, and there may still be a couple more architectural changes to ensure that Symex keeps the tree-sitter layer abstracted, which needs to be done regardless of the actual provider of tree-sitter functionality. I see Simon's work as both the proof-of-concept / reference implementation from a UI standpoint, as well as the means to effect this transition in Symex. As a result, I'd love to merge that work in before exploring long term improvements. This is waiting on me at the moment as I haven't had time to dedicate to it just yet (but, soon! 😓 🙏  ).

Once that's done, the idea of being able to fully abstract the tree-sitter layer is compelling, and tree-edit sounds like it has a really cool approach (e.g. using minikanren and having language-specific customizations to ensure each language has the right behavior - nice, and I'm curious to learn more about it!).

By way of signposting what would be needed for such an integration down the line, I think you've hit it on the head that being able to use tree-edit without the Evil layer would be nice, as Symex uses an evil state already - and of course as it's orthogonal to the core functionality it would help tree-edit be more widely usable. You could keep it all in the tree-edit package and that would be fine, but if you'd like to minimize surface area you could also consider breaking it up into two packages, e.g. tree-edit and tree-edit-evil (similar to what's [planned for Symex](https://github.com/countvajhula/symex.el/issues/26)). Depending on the approach you're taking in your evil layer, it may make sense to combine efforts here, as well.

I will take a look soon and draft a plan for long term tree-sitter integration with tree-edit in mind, at which stage your further input as well as @polaris64 's would be key. Thanks again for stopping by!

**Comment by ethan-leba** [comment]
> To summarize the status of the work here: At the moment, due to the diligent efforts of @polaris64 , proof-of-concept tree-sitter integration is close to ready, and it follows Symex conventions as far as the user experience goes. Part of the work in #33 has been not just about tree-sitter but also making Symex tree-sitter ready, and there may still be a couple more architectural changes to ensure that Symex keeps the tree-sitter layer abstracted, which needs to be done regardless of the actual provider of tree-sitter functionality.

I'm curious to hear more about the abstraction layer here. Are you intending to have identical interfaces and bindings for `symex-ts` and `symex-lisp`? Regardless, having a unified mode for structural editing lisps and nonlisps sounds great! 

> By way of signposting what would be needed for such an integration down the line, I think you've hit it on the head that being able to use tree-edit without the Evil layer would be nice, as Symex uses an evil state already - and of course as it's orthogonal to the core functionality it would help tree-edit be more widely usable. You could keep it all in the tree-edit package and that would be fine, but if you'd like to minimize surface area you could also consider breaking it up into two packages, e.g. tree-edit and tree-edit-evil (similar to what's [planned for Symex](https://github.com/countvajhula/symex.el/issues/26)). Depending on the approach you're taking in your evil layer, it may make sense to combine efforts here, as well.

It's actually already split into two packages (on MELPA, at least). I've just kept them in the same repository for the time being so I don't have to juggle two repos, but I'll probably get around to it at some point.

> I will take a look soon and draft a plan for long term tree-sitter integration with tree-edit in mind, at which stage your further input as well as @polaris64 's would be key. Thanks again for stopping by!

Sounds good!

**Comment by countvajhula** [comment]
> I'm curious to hear more about the abstraction layer here. Are you intending to have identical interfaces and bindings for symex-ts and symex-lisp? Regardless, having a unified mode for structural editing lisps and nonlisps sounds great!

Yes, that's right. The hope is that in general, the features that Symex offers for Lisp should work for non-lisps in the same way. This would be by virtue of being translated through a common language (the symex DSL) on top of the tree abstraction layer.

I'm excited to see what we can do with tree-edit also in the picture 😄

**Comment by polaris64** [comment]
Hi @ethan-leba!

I saw the talk you gave at EmacsConf and I was impressed with the work you've done on tree-edit, nice work!

I agree with @countvajhula that your use of miniKanren to provide language-specific support for doing the right thing when inserting a new node into the tree is very cool. I haven't had chance to take a look at the tree-edit source -- and I admit that I have very little experience with miniKanren and the like -- but am I right in saying that you've created your own language-specific modules to provide this support?

My goal when creating the symex-ts module was to try to abstract the Symex DSL to any language supported by Tree Sitter. As such I tried my hardest not to introduce anything language-specific into that module. It sounds like we could keep the tree navigation code as well as some simple manipulations (e.g. raise, lower, move next/previous) language-agnostic, but for more advanced manipulation (e.g. add a new node inside an function declaration's argument list) we will probably need to be language-specific.

While we're not at this stage just yet, I'm wondering about how your miniKanren inputs are created. Specifically I'm curious as to whether one could be generated from a Tree Sitter grammar? If so then I think we'd be onto something very special, as we could rely solely on the Tree Sitter project to handle both parts when it comes to specific languages! I'm not sure if that's possible, but it's something that has been on my mind so I thought I'd ask :)

@countvajhula: sorry about my lack of input recently, I have unfortunately not had any time to do any more work. Hopefully I'll be able to contribute some more in the new year though!

**Comment by ethan-leba** [comment]
> While we're not at this stage just yet, I'm wondering about how your miniKanren
> inputs are created. Specifically I'm curious as to whether one could be
> generated from a Tree Sitter grammar? If so then I think we'd be onto something
> very special, as we could rely solely on the Tree Sitter project to handle both
> parts when it comes to specific languages! I'm not sure if that's possible, but
> it's something that has been on my mind so I thought I'd ask :)

Tree-sitter dumps a `src/grammar.json` in all of it's repositories, and tree-edit uses this description of the grammar in order to perform it's structural editing operations. So it's very much possible ;)

However Tree-edit does contain some language-specific configuration. 

- When inserting new nodes, it needs to be defined what that new node is. So each node has a template
- Rules about whitespace
- Mapping from keybind to node type
- Some other small things.

There is some work for each new language in that it unearths some assumptions that tree-edit makes that requires code adjustments, but I expect that to keep dropping until we get to a point where adding a new lang is just a matter of configuration.

**Comment by polaris64** [comment]
> There is some work for each new language in that it unearths some assumptions that tree-edit makes that requires code adjustments, but I expect that to keep dropping until we get to a point where adding a new lang is just a matter of configuration.

@ethan-leba thanks for the explanation, that sounds great! I'm definitely looking forward to getting some more work done on this :)

**Comment by anonimitoraf** [comment]
Hi guys, I've personally started to write more and more TSX code recently. I don't mind helping out with the implementation. What's a good place to start looking into?

**Comment by countvajhula** [comment]
@anonimitoraf Nice, welcome! The main immediate goal is to merge the [proof-of-concept tree-sitter work](https://github.com/countvajhula/symex.el/pull/33). After that, exploring integration with tree-edit would be the next step (I haven't had a chance to play with tree-edit yet, but that's the plan :) ).

The description in that PR enumerates the steps before the initial work can be merged. To summarize here, the things that remain to be done are (1) fixing the leap branch feature so that it works for tree-sitter, (2) supporting basic transformations (i.e. not just motions), e.g. `c`, `d`, and `y`, and (3) reviewing the UI and making it consistent between Lisp and non-Lisp.

Of these, (1) is now in code review and should be merged soon. @polaris64 did some work on (2) some time ago but there is no active PR for it, and work on (3) hasn't yet begun.

So, if you can help out with (2) or (3), that would be appreciated. For (2), I would reach out to @polaris64 to get a sense of work that's already there and how you could help. For (3), it's just a matter of checking out the latest development branch (at the moment it's `symex-ts-fix-leap-branch`, but once that's merged it would be the integration branch, `symex-ts-integration`) and playing with it. If you notice any discrepancies with the Symex UX you're used to, then just make a note of it. Once we have a list of these, we can review and see what needs to be done. Since (3) is essentially a UI testing phase, in some respects it is dependent on (2), and it would probably be the most efficient to do them in order, but that's not essential. Any contribution is appreciated!

**Comment by anonimitoraf** [comment]
Sorry for the late reply, I see, at this point I have no clue on the details so thanks for summary. I'll do as you say (i.e. reach out to @polaris64) and try out `symex-ts-fix-leap-branch` + report back

**Comment by polaris64** [comment]
Hi @anonimitoraf,

@countvajhula has just merged `symex-ts-fix-leap-branch` into [`symex-ts-integration`](https://github.com/countvajhula/symex.el/pull/33) (which is the main PR for this work) so that should resolve issue (1) above. I've just started some work on (2) so I'll create a PR for it as soon as there's something to show and I'll be sure to let you know about it!

**Comment by anonimitoraf** [comment]
Sweet, thanks! I've started trying it out. I've noticed intermittent lags with slurping/barfing

Here's a sample profiler report
```

6207  92% - command-execute
5656  84%  - funcall-interactively
2540  37%   - hydra-symex/symex-capture-forward
2540  37%    - condition-case
2540  37%     - progn
2540  37%      - hydra--call-interactively-remap-maybe
2540  37%       - call-interactively
2540  37%        - funcall-interactively
2540  37%         - symex-capture-forward
2540  37%          - symex--capture-forward
2540  37%           - lispy-forward-slurp-sexp
2531  37%            - lispy-slurp
2517  37%             - lispy--reindent
2507  37%              - indent-sexp
1677  24%               - indent-for-comment
1657  24%                - indent-according-to-mode
1650  24%                 - typescript-indent-line
1428  21%                  - typescript--proper-indentation
 814  12%                   - typescript--backward-to-parameter-list
   3   0%                    - backward-sexp
   3   0%                       forward-sexp
 578   8%                   + typescript--ctrl-statement-indentation
  24   0%                   + typescript--continued-expression-p
 203   3%                  + indent-line-to
   3   0%                    syntax-ppss
  17   0%                + comment-search-forward
 587   8%               + indent-line-to
 183   2%               + lisp-indent-calc-next
   7   0%               + forward-sexp
   6   0%              + backward-list
  14   0%             + lispy--slurp-forward
   9   0%            + lispy--out-forward
1947  29%   - hydra-symex/symex-emit-backward
1947  29%    - condition-case
1947  29%     - progn
1947  29%      - hydra--call-interactively-remap-maybe
1947  29%       - call-interactively
1947  29%        + funcall-interactively
1166  17%   + hydra-symex/symex-capture-backward
   3   0%   + evil-force-normal-state
 551   8%  + byte-code
 328   4% + ...
  99   1% + redisplay_internal (C function)
  25   0% + timer-event-handler
  16   0% + #<compiled 0xe8d77218f769fed>
  14   0% + tide-net-filter
   7   0% + #<compiled -0x4102a58c4b7b1bb>
   7   0% + jit-lock--antiblink-post-command
   4   0% + evil-repeat-post-hook
   3   0%   flycheck-error-list-update-source
```

**Comment by countvajhula** [comment]
@anonimitoraf There [haven't been any changes](https://github.com/countvajhula/symex.el/pull/33/files#diff-3ac173513701ddffd6c4cecac617bd9152ec5b67983040614aa77073cb980042) in emit/capture ("barf/slurp") in this PR, so that's surprising, but maybe there's something else going on. Some questions:

- is it intermittent, or does it always happen?
- does it always happen when you do a specific thing?
- could you try using the master branch and seeing if it still happens?
- any way to reproduce it consistently would be helpful

Also, that performance profile is cool! How did you generate it, out of curiosity?

**Comment by polaris64** [comment]
> Also, that performance profile is cool! How did you generate it, out of curiosity?

A very handy feature of Emacs: `profiler-start`, `profiler-stop` and `profiler-report` :)

**Comment by anonimitoraf** [comment]
Ah, interesting.
* Yes, intermittent. Only barfs/slurps are laggy
* Hmm, I don't think so.
* I'll try later (on some Lisp code, right?)

**Comment by tommy-mor** [comment]
a small issue i've noticed with some actions (particularly ``o``,  ``symex-open-line-after``) don't act how I would expect.
normal symex

1. normal symex (correct behavior)
   before ![Screenshot from 2022-06-18 15-22-30](https://user-images.githubusercontent.com/5810139/174454521-fc7d56d9-9bd0-42f3-9233-a3b027214dcd.png)
   after- ![Screenshot from 2022-06-18 15-22-37](https://user-images.githubusercontent.com/5810139/174454514-2214ffcc-e46b-42d7-98dd-f35db96732c1.png)

2. symex tree sitter (wrong)
  before ![Screenshot from 2022-06-18 15-22-57](https://user-images.githubusercontent.com/5810139/174454510-5723af74-f981-485c-8e9a-85645c104575.png)
  after- ![Screenshot from 2022-06-18 15-23-02](https://user-images.githubusercontent.com/5810139/174454503-6a1a8fee-1c8e-4353-bfc1-97d8e2e6c469.png)


it seems like symex-open-line-after is opening a line after not the currently selected node, but its first child. If I press
``symex-go-up`` from 2.before, it selects the first child node, which seems to be where its putting the newline for correctly.

![Screenshot from 2022-06-18 15-43-07](https://user-images.githubusercontent.com/5810139/174454977-8e960541-cdbd-4301-a7c0-ddd4a58fc3cc.png)

**Comment by countvajhula** [comment]
At the moment the branch only contains "motions" for tree sitter. The PR that Simon is putting together will add some basic transformations. I was thinking just `c` `d` and `y` and `p` to start, but it might make sense to add a few others that are common and aren't too involved - e.g. `o` and `O` might be good candidates. Probably the user testing phase (which you have begun 😄 ) would make clear what features are expected.

For signposting purposes, I expect we'll be able to rely on tree-edit for a lot of functionality down the line - so the purpose of the current work is mostly to prove the concept by providing a basic, useful Symex experience for tree-sitter, rather than to provide all of the existing Lisp features.

**Comment by tommy-mor** [comment]
I see, I guess I misunderstood because the action still kind of worked... its tantalizingly close to being usable :)

**Comment by polaris64** [comment]
Hi all, I've just [created a draft PR](https://github.com/countvajhula/symex.el/pull/47) for this new tree sitter transformation support. It's currently a draft as I've only implemented some of the transformations so far; please see the PR for a full list of what has been done and what's left to do.

@anonimitoraf please see this PR for the aforementioned step (2).

@tommy-mor just for you, I implemented `symex-open-line-after` and `symex-open-line-before` :)

Please leave a comment on the draft PR if you've tested these changes and hopefully we can get it merged soon!

**Comment by devcarbon-com** [comment]
Where does this currently stand?

Is there of todo's?

**Comment by countvajhula** [comment]
@devcarbon-com Slow and steady progress. The latest updates are on the [integration PR](https://github.com/drym-org/symex.el/pull/33) for this work, specifically the last [two](https://github.com/drym-org/symex.el/pull/33#issuecomment-1232288593) [comments](https://github.com/drym-org/symex.el/pull/33#issuecomment-1232325656). There is a Kanban board with the remaining items linked in the PR description. Personally, I've been focusing on my recent [EmacsConf talk](https://emacsconf.org/2022/talks/maint/) and haven't had a lot of time to continue this work, but it's still very much on the cards. Contributions are of course welcome 🙂

**Comment by countvajhula** [comment]
There is now "alpha" support for this, as of #33 . This is still not "officially" supported, but full support is aimed to be completed by **March 20, 2023** (the first day of Spring, when squirrels come out of their nests and hollows to play in the sun again 🐭 🌷 🌤️ ) for when we're targeting a 2.0 release. Any feedback on the alpha release is appreciated if you are feeling brave, but of course, please expect things to be broken!
